src/core/toga/widgets/textinput.py

```diff
@@ -100,10 +106,40 @@ def on_change(self):
 
     @on_change.setter
     def on_change(self, handler):
-        """Set the handler to invoke when the value is changeed.
+        """Set the handler to invoke when the value is changed.
 
         Args:
-            handler (:obj:`callable`): The handler to invoke when the value is changeed.
+            handler (:obj:`callable`): The handler to invoke when the value is changed.
         """
         self._on_change = wrapped_handler(self, handler)
         self._impl.set_on_change(self._on_change)
+
+    @property
+    def validator(self):
+        return self._validator
+
+    @validator.setter
+    def validator(self, validator):
+        self._validator = validator
+        self.validate()
+
+    def validate(self):
+        if self.validator is None:
+            error_message = None
+        else:
+            error_message = self.validator(self.value)
+        if error_message is None:
+            self.unset_error()
+        else:
+            self.set_error(error_message)
+
+    def is_valid(self):
+        if self.validator is None:
+            return True
+        return self.validator(self.value) is None
+
+    def set_error(self, error_message):
+        self._impl.set_error(error_message)
+
+    def unset_error(self):
```

Comment:
I don't think `set_error` and `unset_error` should be exposed as public API. AFAICT, they're entirely internal mechanisms, and are only needed by `validate()`. 

I'd be inclined to remove these these entirely, and just invoke `self._impl.set_error()` and `self._impl.unset_error()` as required inside validate. I can't think of any reason they'd need to be overridden at the *interface* level.